File: my-folder/0040-combination-sum-ii/solution.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
       List<List <Integer>> res = new ArrayList<>();
       Arrays.sort(candidates);
       backtrack(candidates, target, res, new ArrayList<>(), 0);
       return res;
    }

    private void backtrack (int[] candidates, int target, List<List<Integer>> res, List<Integer> temp, int start) {
        if (target == 0) {
            res.add(new ArrayList<>(temp));
            return;
        }
        if (target < 0) return;
        for (int i = start; i < candidates.length; i++) {
            if (i > start && candidates[i] == candidates[i-1]) continue;
            temp.add(candidates[i]);
            backtrack(candidates, target - candidates[i], res, temp, i+1);
            temp.remove((temp.size() - 1));
        }
    }
}

File: my-folder/0040-combination-sum-ii/solution.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        candidates.sort()
        def backtrack(candidates, target, start, temp, res):
            if target < 0: return
            if target == 0: res.append(temp)
            for i in range(start, len(candidates)):
                if i > start and candidates[i] == candidates[i-1]:
                    continue
                backtrack(candidates, target - candidates[i], i+1, temp+[candidates[i]], res)
            
        backtrack(candidates, target, 0, [], res)
        return res

